File: python-algorithm/ds_9_bit_operation/L260-m.cpp
//
// Created by 李恒 on 2020/6/29.
//

//两个不相等的元素在位级表示上必定会有一位存在不同。
//将数组的所有元素异或得到的结果为不存在重复的两个元素异或的结果。
//diff &= -diff 得到出 diff 最右侧不为 0 的位，也就是不存在重复的两个元素在位级表示上最右侧不同的那一位，利用这一位就可以将两个元素区分开来。

#include <iostream>
#include <vector>
using namespace std;
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int diff = 0;
        for (int num : nums) diff ^= num;
        diff &= -diff;  // 得到最右一位不为0的位？？？还没懂
        vector<int> ret(2,0);
        for (int num : nums) {
            if ((num & diff) == 0)
                ret[0] ^= num;
            else
                ret[1] ^= num;
        }
        return ret;
    }

};

File: python-algorithm/ds_9_bit_operation/L260-m.py
# -*- coding: utf-8 -*-


class Solution:
    def singleNumber(self, nums: int) -> List[int]:
        # difference between two numbers (x and y) which were seen only once
        bitmask = 0
        # 异或，相同为0，不同为1
        for num in nums:
            bitmask ^= num
        # bitmask是2个只出现1次的数的异或结果，代表了2者二进制位的差异

        # rightmost 1-bit diff between x and y
        diff = bitmask & (-bitmask)  # 可以保留bitmask最右边为1的位

        one_of_them = 0
        for num in nums:
            # bitmask which will contain only x
            # 通过循环排除掉出现2次的数
            if num & diff:
                one_of_them ^= num
        return [one_of_them, bitmask ^ one_of_them]

